dbcon/joblist/primitivestep.h

```diff
@@ -1421,7 +1421,8 @@ class TupleBPS : public BatchPrimitive, public TupleDeliveryStep
     bool BPPIsAllocated;
     uint32_t uniqueID;
     ResourceManager* fRm;
-
+    boost::shared_ptr<int64_t> fJoinMemLimit;
```

Comment:
It seems that it would be better to introduce a new RM::returnMemory overload b/c shared_ptr is an overkill for a simple atomic counter.

Reply:
There's a shared pointer called fSessionMemLimit that is used in calls to
fRm->getMemory(newSize, fSessionMemLimit)
fSessionMemLimit is optionally set by the .cnf variable columnstore_um_mem_limit. It's passed around as a shared ptr so various parts of the query can access it. It controls a potentially smaller amount of memory that may trigger "UM limit exceeded" for a single session while still maintaining the larger xml setting for overall memory limit.

The reason that it's a shared pointer is that getMemory decrements it and we want all threads to see the decrement and similarly with releaseMemory.

In the case here, I want to ignore fSessionMemLimit, but the getMemory function demands a variable (shared ptr), so I set fJoinMemLimit to the larger um mem limit and and pass that.

Reply:
Thank you for the detailed explanation.
JFYI This approach do atomic ops twice that is why I don't like it. First shared_ptr atomically increments its internal counter and secondly RM atomically increments/decrements the value. It would be better to just do it once using std::atomic variable using memory_order::acquire/release modes IMHO.

Reply:
by changing the getMemory function to (adding '&')
`bool ResourceManager::getMemory(int64_t amount, boost::shared_ptr<int64_t>& sessionLimit, bool patience)
`
It should remove the shared_ptr increment.